spring-cloud-gcp-pubsub/src/main/java/com/google/cloud/spring/pubsub/core/PubSubConfiguration.java

```diff
@@ -45,26 +50,34 @@ public Publisher getPublisher() {
 		return this.publisher;
 	}
 
+	public Map<String, PubSubProperties> getProperties() {
+		return this.properties;
+	}
+
+	public Subscriber getSubscriber(String name) {
+		PubSubProperties property = new PubSubProperties();
+		property.setSubscriber(this.subscriber);
+		this.properties.putIfAbsent(name, property);
+		return this.properties.get(name).getSubscriber();
```

Comment:
I would immediately return the object if it's found in the map first. Then construct `PubSubProperties` only if we know there was not one at that particular point in time. Otherwise we'll end up creating a bunch of `PubSubProperties` objects and leaving them to immediately be garbage collected.

`putIfAbsent` is still necessary to prevent a race condition, in which the properties were not in the map at the time of the initial check, but have been added by another thread since then.

The current order also makes it look like the method creates `PubSubProperties`, sets a subscriber field, and then returns the subscriber field... So it causes a question why not just return `this.subscriber` here. And *then* you notice the _putIfAbsent_, resolving that question.

Reply:
Ah I missed this, thank you